### ISSUE TITLE
[Knative] Failed to run lithops test

### ISSUE THREAD

**Issue by zhanggbj** [lithops-cloud/lithops#406]
I can run lithop successfully with config https://github.com/lithops-cloud/lithops/issues/391 with a CodeEngine accout, which means I have set kubeconfig only have limited access to the platform.

Then I changed Kubeconfig which have full acess to the CodeEngine, that lithops test failed with following error, any suggestions? 

I can see the Knative service is created successfully, and looks like lithops failed at invoker?
```
# curl http://kserving--lithops-knative-v36--221dev1027--256mb.default.dev-serving.codeengine.dev.appdomain.cloud/preinstalls -L
{"preinstalls":[["__future__",false],["_asyncio",false],["_bisect",false],["_blake2",false],["_bootlocale",false],["_bz2",false],["_codecs_cn",false],["_codecs_hk",false],["_codecs_iso2022",false],["_codecs_jp",false],["_codecs_kr",false],["_codecs_tw",false],["_collections_abc",false],["_compat_pickle",false],["_compression",false],["_crypt",false],["_csv",false],["_ctypes",false],["_ctypes_test",false],["_curses",false],["_curses_panel",false],["_datetime",false],["_decimal",false],["_distutils_hack",true],["_dummy_thread",false],["_elementtree",false],["_gdbm",false],["_hashlib",false],["_heapq",false],["_json",false],["_lsprof",false],["_lzma",false],["_markupbase",false],["_md5",false],["_multibytecodec",false],["_multiprocessing",false],["_opcode",false],["_osx_support",false],["_pickle",false],["_posixsubprocess",false],["_pydecimal",false],["_pyio",false],["_random",false],["_sha1",false],["_sha256",false],["_sha3",false],["_sha512",false],["_sitebuiltins",false],["_socket",false],["_sqlite3",false],["_ssl",false],["_strptime",false],["_struct",false],["_sysconfigdata_m_linux_x86_64-linux-gnu",false],["_testbuffer",false],["_testcapi",false],["_testimportmultiple",false],["_testmultiphase",false],["_threading_local",false],["_tkinter",false],["_weakrefset",false],["abc",false],["aifc",false],["antigravity",false],["argparse",false],["array",false],["ast",false],["asynchat",false],["asyncio",true],["asyncore",false],["audioop",false],["base64",false],["bdb",false],["binascii",false],["binhex",false],["bisect",false],["bz2",false],["cProfile",false],["cachetools",true],["calendar",false],["certifi",true],["cgi",false],["cgitb",false],["chardet",true],["chunk",false],["click",true],["cmath",false],["cmd",false],["code",false],["codecs",false],["codeop",false],["collections",true],["colorsys",false],["compileall",false],["concurrent",true],["configparser",false],["contextlib",false],["copy",false],["copyreg",false],["crypt",false],["csv",false],["ctypes",true],["curses",true],["datetime",false],["dateutil",true],["dbm",true],["decimal",false],["difflib",false],["dis",false],["distutils",true],["doctest",false],["docutils",true],["dummy_threading",false],["easy_install",false],["email",true],["encodings",true],["ensurepip",true],["enum",false],["fcntl",false],["filecmp",false],["fileinput",false],["flask",true],["fnmatch",false],["formatter",false],["fractions",false],["ftplib",false],["functools",false],["genericpath",false],["getopt",false],["getpass",false],["gettext",false],["gevent",true],["glob",false],["glob2",true],["greenlet",false],["grp",false],["gunicorn",true],["gzip",false],["hashlib",false],["heapq",false],["hmac",false],["html",true],["http",true],["ibm_boto3",true],["ibm_botocore",true],["ibm_s3transfer",true],["idlelib",true],["idna",true],["imaplib",false],["imghdr",false],["imp",false],["importlib",true],["inspect",false],["io",false],["ipaddress",false],["itsdangerous",true],["jinja2",true],["jmespath",true],["jp",false],["json",true],["keyword",false],["kubernetes",true],["lib2to3",true],["linecache",false],["lithops",true],["lithopsproxy",false],["locale",false],["logging",true],["lzma",false],["macpath",false],["macurl2path",false],["mailbox",false],["mailcap",false],["markupsafe",true],["math",false],["mimetypes",false],["mmap",false],["modulefinder",false],["multiprocessing",true],["netrc",false],["nis",false],["nntplib",false],["ntpath",false],["nturl2path",false],["numbers",false],["numpy",true],["oauthlib",true],["opcode",false],["operator",false],["optparse",false],["os",false],["ossaudiodev",false],["parser",false],["pathlib",false],["pdb",false],["pickle",false],["pickletools",false],["pika",true],["pip",true],["pipes",false],["pkg_resources",true],["pkgutil",false],["platform",false],["plistlib",false],["poplib",false],["posixpath",false],["pprint",false],["profile",false],["pstats",false],["pty",false],["py_compile",false],["pyasn1",true],["pyasn1_modules",true],["pyclbr",false],["pydoc",false],["pydoc_data",true],["pyexpat",false],["queue",false],["quopri",false],["random",false],["re",false],["readline",false],["redis",true],["reprlib",false],["requests",true],["requests_oauthlib",true],["resource",false],["rlcompleter",false],["rsa",true],["rst2html",false],["rst2html4",false],["rst2html5",false],["rst2latex",false],["rst2man",false],["rst2odt",false],["rst2odt_prepstyles",false],["rst2pseudoxml",false],["rst2s5",false],["rst2xetex",false],["rst2xml",false],["rstpep2html",false],["runpy",false],["sched",false],["secrets",false],["select",false],["selectors",false],["setuptools",true],["shelve",false],["shlex",false],["shutil",false],["signal",false],["site",false],["six",false],["smtpd",false],["smtplib",false],["sndhdr",false],["socket",false],["socketserver",false],["spwd",false],["sqlite3",true],["sre_compile",false],["sre_constants",false],["sre_parse",false],["ssl",false],["stat",false],["statistics",false],["string",false],["stringprep",false],["struct",false],["subprocess",false],["sunau",false],["symbol",false],["symtable",false],["sysconfig",false],["syslog",false],["tabnanny",false],["tarfile",false],["telnetlib",false],["tempfile",false],["termios",false],["textwrap",false],["this",false],["threading",false],["timeit",false],["tkinter",true],["token",false],["tokenize",false],["trace",false],["traceback",false],["tracemalloc",false],["tty",false],["turtle",false],["turtledemo",true],["types",false],["typing",false],["unicodedata",false],["unittest",true],["urllib",true],["urllib3",true],["uu",false],["uuid",false],["venv",true],["warnings",false],["wave",false],["weakref",false],["webbrowser",false],["websocket",true],["werkzeug",true],["wheel",true],["wsdump",false],["wsgiref",true],["xdrlib",false],["xml",true],["xmlrpc",true],["xxlimited",false],["yaml",true],["zipapp",false],["zipfile",false],["zlib",false]],"python_ver":"3.6"}
```

```
$ lithops test
DEBUG:lithops.serverless.backends.knative.knative:Creating Lithops runtime service in Knative
DEBUG:lithops.serverless.backends.knative.knative:Service name: kserving--lithops-knative-v36--221dev1027--256mb
DEBUG:lithops.serverless.backends.knative.knative:Namespace: default
DEBUG:lithops.serverless.backends.knative.knative:Runtime Service created - URL: http://kserving--lithops-knative-v36--221dev1027--256mb.default.dev-serving.codeengine.dev.appdomain.cloud
DEBUG:lithops.serverless.backends.knative.knative:Extracting Python modules list from: kserving/lithops-knative-v36:221dev1027
Traceback (most recent call last):
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 444, in _generate_runtime_meta
    runtime_meta = self.invoke(docker_image_name, memory, payload, return_result=True)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 615, in invoke
    raise e
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 596, in invoke
    conn.request("POST", route, body=json.dumps(payload), headers=headers)
  File "/usr/lib/python3.6/http/client.py", line 1281, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1327, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1276, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1042, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 980, in send
    self.connect()
  File "/usr/lib/python3.6/http/client.py", line 952, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/usr/lib/python3.6/socket.py", line 704, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.6/socket.py", line 745, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/lithops", line 33, in <module>
    sys.exit(load_entry_point('lithops', 'console_scripts', 'lithops')())
  File "/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/root/workspace/flops/lithops/lithops/cli/cli.py", line 34, in test_function
    pw.call_async(hello, 'World')
  File "/root/workspace/flops/lithops/lithops/executors.py", line 156, in call_async
    runtime_meta = self.invoker.select_runtime(job_id, runtime_memory)
  File "/root/workspace/flops/lithops/lithops/invokers.py", line 117, in select_runtime
    runtime_meta = self.compute_handler.create_runtime(runtime_name, runtime_memory, timeout=timeout)
  File "/root/workspace/flops/lithops/lithops/serverless/serverless.py", line 62, in create_runtime
    return self.backend.create_runtime(runtime_name, memory, timeout=timeout)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 468, in create_runtime
    runtime_meta = self._generate_runtime_meta(docker_image_name, memory)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 446, in _generate_runtime_meta
    raise Exception("Unable to extract the preinstalled modules from the runtime: {}".format(e))
Exception: Unable to extract the preinstalled modules from the runtime: [Errno -2] Name or service not known
```

**Comment by zhanggbj** [comment]
Hi @gilv @JosepSampe would you please help to take a look? Thanks!

**Comment by JosepSampe** [comment]
@zhanggbj Seems it is working correctly from my side, so I suspect it is a "specific path issue", I mean, you have followed some steps that led to an issue.
Can you try cleaning lithops by:
```
lithops clean
```
and then try again:
```
lithops test
```

`lithops clean` functionality cleans everything related to lithops except the configuration, so its like you start from scratch without any possible interference of deployed runtimes or cache.

**Comment by zhanggbj** [comment]
@zhanggbj 
I did a cleanup but get the same error, maybe one question is there any difference b/w a CodeEngine Project or a normal Knative deployment?  Thanks!
```
[root@zhanggbjwsp flops (⎈ |dev-serving-s02:default)]# lithops clean
INFO:lithops.cli.utils:Cleaning all Lithops information
INFO:lithops.config:Lithops v2.2.1dev1027
DEBUG:lithops.config:Loading configuration
INFO:lithops.config:Getting configuration from /root/workspace/flops/config
DEBUG:lithops.config:Loading Serverless backend module: knative
DEBUG:lithops.config:Loading Storage backend module: ibm_cos
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:Creating IBM COS client
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:Set IBM COS Endpoint to https://s3.us-south.cloud-object-storage.appdomain.cloud
DEBUG:lithops.libs.ibm_iam.ibm_iam:Using IBM IAM API Key - Reusing Token from local cache
DEBUG:lithops.libs.ibm_iam.ibm_iam:Token expiry time: 2020-10-27 06:15:30.139428-04:00 - Minutes left: 35
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:IBM COS client created successfully
DEBUG:lithops.serverless.backends.knative.knative:Loaded service host suffix: .default.dev-serving.codeengine.dev.appdomain.cloud
INFO:lithops.serverless.backends.knative.knative:Lithops v2.2.1dev1027 init for Knative - Istio Endpoint: http://None:30030
INFO:lithops.serverless.backends.knative.knative:Deleting runtime: kserving--lithops-knative-v36--221dev1027--256mb
DEBUG:lithops.storage.utils:Going to delete all objects from bucket 'lithops-serving' and prefix 'lithops.runtimes'
DEBUG:lithops.storage.utils:Finished deleting objects, total found: 0
DEBUG:lithops.storage.utils:Going to delete all objects from bucket 'lithops-serving' and prefix 'lithops.jobs'
DEBUG:lithops.storage.utils:Finished deleting objects, total found: 0
```

**Comment by JosepSampe** [comment]
@zhanggbj Can you post full DEBUG logs from `lithops test`?

**Comment by zhanggbj** [comment]
@JosepSampe here it is

```
INFO:lithops.config:Lithops v2.2.1dev1027
DEBUG:lithops.config:Loading configuration
INFO:lithops.config:Getting configuration from /root/workspace/flops/config
DEBUG:lithops.config:Loading Serverless backend module: knative
DEBUG:lithops.config:Loading Storage backend module: ibm_cos
INFO:lithops.config:Lithops v2.2.1dev1027
DEBUG:lithops.config:Loading configuration
DEBUG:lithops.config:Loading Serverless backend module: knative
DEBUG:lithops.config:Loading Storage backend module: ibm_cos
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:Creating IBM COS client
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:Set IBM COS Endpoint to https://s3.us-south.cloud-object-storage.appdomain.cloud
DEBUG:lithops.libs.ibm_iam.ibm_iam:Using IBM IAM API Key - Token expired. Requesting new token
DEBUG:lithops.storage.backends.ibm_cos.ibm_cos:IBM COS client created successfully
DEBUG:lithops.serverless.backends.knative.knative:Loaded service host suffix: None
INFO:lithops.serverless.backends.knative.knative:Lithops v2.2.1dev1027 init for Knative - Istio Endpoint: http://None:30030
DEBUG:lithops.invokers:ExecutorID 11ed3c/0 - Total available workers: 100
DEBUG:lithops.invokers:ExecutorID 11ed3c/0 - Serverless invoker created
INFO:lithops.executors:Serverless Executor created with ID: 11ed3c/0
INFO:lithops.invokers:ExecutorID 11ed3c/0 | JobID A000 - Selected Runtime: kserving/lithops-knative-v36:221dev1027 - 256MB
DEBUG:lithops.storage.storage:Runtime metadata not found in local cache. Retrieving it from storage
DEBUG:lithops.storage.storage:Trying to download runtime metadata from: ibm_cos://lithops-serving/lithops.runtimes/2.2.1dev1027/dev-serving-s02/btuv3qes0snr5q9jun4g/default/kserving--lithops-knative-v36--221dev1027--256mb.meta.json
DEBUG:lithops.storage.storage:Runtime metadata not found in storage
DEBUG:lithops.invokers:ExecutorID 11ed3c/0 | JobID A000 - Runtime kserving/lithops-knative-v36:221dev1027 with 256MB is not yet installed
DEBUG:lithops.invokers:Creating runtime: kserving/lithops-knative-v36:221dev1027, memory: 256MB
INFO:lithops.serverless.backends.knative.knative:Building a new docker image from Dockerfile
INFO:lithops.serverless.backends.knative.knative:Docker image name: kserving/lithops-knative-v36:221dev1027
DEBUG:lithops.serverless.utils:Creating function handler zip in /root/workspace/flops/lithops_knative.zip
Sending build context to Docker daemon  58.88MB
Step 1/8 : FROM python:3.6-slim-buster
 ---> f9fcc0353bbd
Step 2/8 : RUN apt-get update && apt-get install -y         zip         && rm -rf /var/lib/apt/lists/*
 ---> Using cache
 ---> 9c0b6ad71f33
Step 3/8 : RUN pip install --upgrade setuptools six pip     && pip install --no-cache-dir         gunicorn==19.9.0         pika==0.13.1         flask         gevent         glob2         ibm-cos-sdk         redis         requests         PyYAML         kubernetes         numpy
 ---> Using cache
 ---> f5d56096f2b9
Step 4/8 : ENV APP_HOME /lithops
 ---> Using cache
 ---> 5f13db7cab21
Step 5/8 : WORKDIR $APP_HOME
 ---> Using cache
 ---> a049373d7085
Step 6/8 : COPY lithops_knative.zip .
 ---> Using cache
 ---> 9e6449a9e496
Step 7/8 : RUN unzip lithops_knative.zip && rm lithops_knative.zip
 ---> Using cache
 ---> 4a1018ee8576
Step 8/8 : CMD exec gunicorn --bind :$PORT --workers 1 --timeout 600 lithopsproxy:proxy
 ---> Using cache
 ---> 39201bf6003a
Successfully built 39201bf6003a
Successfully tagged kserving/lithops-knative-v36:221dev1027
The push refers to repository [docker.io/kserving/lithops-knative-v36]
7673615821e0: Layer already exists
762b4bd8473c: Layer already exists
5229f161a2d0: Layer already exists
8723d030b279: Layer already exists
8969d0d592f3: Layer already exists
c04dd0e99e62: Layer already exists
2aa642c0a1e4: Layer already exists
bbc6d16a112a: Layer already exists
225ef82ca30a: Layer already exists
d0fe97fa8b8c: Layer already exists
221dev1027: digest: sha256:5cc605d6dc12d273945a8fe5f7f2980ffa718823f1ee836616a1892ab2ed600c size: 2419
DEBUG:lithops.serverless.backends.knative.knative:Creating Lithops runtime service in Knative
DEBUG:lithops.serverless.backends.knative.knative:Service name: kserving--lithops-knative-v36--221dev1027--256mb
DEBUG:lithops.serverless.backends.knative.knative:Namespace: default
DEBUG:lithops.serverless.backends.knative.knative:Runtime Service created - URL: http://kserving--lithops-knative-v36--221dev1027--256mb.default.dev-serving.codeengine.dev.appdomain.cloud
DEBUG:lithops.serverless.backends.knative.knative:Extracting Python modules list from: kserving/lithops-knative-v36:221dev1027
Traceback (most recent call last):
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 444, in _generate_runtime_meta
    runtime_meta = self.invoke(docker_image_name, memory, payload, return_result=True)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 614, in invoke
    raise e
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 596, in invoke
    conn.request("POST", route, body=json.dumps(payload), headers=headers)
  File "/usr/lib/python3.6/http/client.py", line 1281, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1327, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1276, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1042, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 980, in send
    self.connect()
  File "/usr/lib/python3.6/http/client.py", line 952, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/usr/lib/python3.6/socket.py", line 704, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.6/socket.py", line 745, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/lithops", line 33, in <module>
    sys.exit(load_entry_point('lithops', 'console_scripts', 'lithops')())
  File "/usr/lib/python3/dist-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/root/workspace/flops/lithops/lithops/cli/cli.py", line 34, in test_function
    pw.call_async(hello, 'World')
  File "/root/workspace/flops/lithops/lithops/executors.py", line 156, in call_async
    runtime_meta = self.invoker.select_runtime(job_id, runtime_memory)
  File "/root/workspace/flops/lithops/lithops/invokers.py", line 117, in select_runtime
    runtime_meta = self.compute_handler.create_runtime(runtime_name, runtime_memory, timeout=timeout)
  File "/root/workspace/flops/lithops/lithops/serverless/serverless.py", line 62, in create_runtime
    return self.backend.create_runtime(runtime_name, memory, timeout=timeout)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 468, in create_runtime
    runtime_meta = self._generate_runtime_meta(docker_image_name, memory)
  File "/root/workspace/flops/lithops/lithops/serverless/backends/knative/knative.py", line 446, in _generate_runtime_meta
    raise Exception("Unable to extract the preinstalled modules from the runtime: {}".format(e))
Exception: Unable to extract the preinstalled modules from the runtime: [Errno -2] Name or service not known
```

**Comment by JosepSampe** [comment]
@zhanggbj I can see one line that should not appear, and I think it might cause the issue:
```
INFO:lithops.serverless.backends.knative.knative:Lithops v2.2.1dev1027 init for Knative - Istio Endpoint: http://None:30030
```
I think your new kubeconfig has extra rights for accessing to the platform, and lithops follows a path that we never faced.
I'm adding a fix, thanks

**Comment by JosepSampe** [comment]
@zhanggbj Added a tentative fix in #409 . Can you test master branch?

**Comment by zhanggbj** [comment]
@JosepSampe Thanks!! 👍  It works!
